### PR TITLE
Password reset and invitations backend

### PR DESCRIPTION
**Pull request by mike12345567**
## Description
This purely implements the backend for these features as I made a number of changes and updated a fair number of tests, felt it was best to get the backend merged before developing any usage of it in the frontend.

Also, noting that about 2-2.5K of the + changes are from yarn locks, as during all the merges with next these got quite out of sync, refreshed them to fix.

This PR introduces a few things:
1. In the server when testing bull is no longer used, the `inMemoryQueue` is instead used. This is simply because we haven't built a system to run Redis/CouchDB as part of our CI and its a lot easier to just run redis is memory (as the integration doesn't hugely need tested).
2. The auth library introduces a redis API, which will be in memory when running in tests. The redis API has a few features, but it namely handles all of the connection, disconnection, refreshing connection, handling objects and exposing functions like the ability to scan. It also makes use of a "database" system where all keys are prefixed (similar to how we use Couch) meaning that we can create different fragments of data inside a single Redis deployment. This is all built atop of `ioredis`.
3. I've password protected the redis instance now as part of the self hosted configuration as previously it was simply publically exposed.
4. Redis doesn't pass cleanly through Envoy for some reason, the requests aren't entirely http based so it seems to mess things up. For now in development I've made it so that it goes directly to a docker exposed port - this isn't important when actually deployed as the redis instance should only be communicated with via direct connection within the docker composed network - if we want to cluster redis will load balance itself naturally.
5. The new features have a basic suite of tests just to confirm the whole flows, as part of this I did a little rework of the worker test suite and included it in the main CI runs.

As far as the API of these new features goes, there are the following endpoints:
1. POST `/api/admin/users/invite` - invites a user if the email address provided is not in use. Right now this sends a link in the email which contains a code which is valid for a day (TTL expiry in Redis) however how we make use of this link is really dependent on the implementation in the frontend (and routing).
2. POST `/api/admin/users/invite/accept` - accept an invitation, basically when a user clicks a link they received in the email (from above endpoint) they'll be taken to the frontend and asked to provide their user details (right now just their password) the link also provides the `code` as a query string parameter, which can be provided to the `inviteCode` of this API.
3. POST `/api/admin/auth/reset` - very similar to the invite API, simply sends an email with a link that contains a code to reset password.
4. POST `/api/admin/auth/reset/update` - as with the accept invite this simply takes the code provided in the `resetCode` and a password, which then updates the users password (to the provided one).

### CODE REVIEW

File: packages/worker/src/api/controllers/admin/configs.js [hunk 1]
```diff
@@ -16,7 +16,7 @@ exports.save = async function (ctx) {
 
   // Config does not exist yet
   if (!ctx.request.body._id) {
-    ctx.request.body._id = generateConfigID({
+    config._id = generateConfigID({
```
Comment (shogunpurple) [inline packages/worker/src/api/controllers/admin/configs.js:19]: When I was doing the OAuth stuff, I updated the structure of this so that it matches the joi validation. The config docs were being sent in a different format than they were being saved, and the actual configuration fields were alongside the rest of the fields in the document. The shape of the doc is now like this:

```
{
  _id: 123,
 type: "google",
 config: {
   something: 123
 }
}
```

**Review by shogunpurple — APPROVED** [review]
This is awesome work - really like the foundational elements around redis and temporary record generation for us going forward.

Good catch on fixing bull with tests etc - I forgot to re-enable the in memory queue for tests!

File: packages/worker/src/api/controllers/admin/configs.js (hunk 1 above)
Comment (mike12345567) [reply packages/worker/src/api/controllers/admin/configs.js:19]: Yes, I was thinking about this and how to manage it, for Joi to do conditionals it had to have the type somewhere in the structure and the config as part of a structure we could have alternatives for (e.g. an array or object).

I was thinking that the best way to manage this going forward rather than requiring every read of the config do something like `const { config } = await determineScopedConfig(...)` we could just put everything into the actual config structure, I'll move the type into it too (I didn't originally but I think it should be in there too) so that the structure that is saved is like:
```
{
  _id: 123,
 type: "google",
 something: 123
}
```

basically doing `ctx.request.body = { ...ctx.request.body.config, ...ctx.request.body}; delete ctx.request.body.config`.
